sqlx-postgres/src/options/mod.rs

```diff
@@ -81,15 +82,40 @@ impl PgConnectOptions {
             // `-----BEGIN CERTIFICATE-----` and so will not attempt to parse
             // a PEM-encoded private key.
             ssl_client_key: var("PGSSLKEY").ok().map(CertificateInput::from),
-            ssl_mode: var("PGSSLMODE")
-                .ok()
-                .and_then(|v| v.parse().ok())
-                .unwrap_or_default(),
-            statement_cache_capacity: 100,
             application_name: var("PGAPPNAME").ok(),
+            options: var("PGOPTIONS").ok(),
+            ..Self::new_without_environment(host, port, username, ssl_mode)
+        }
+    }
+
+    /// Create a minimal set of connection options _without_ applying any environment.
+    ///
+    /// To be used in situations the environment is not controlled enough to be relied upon.
+    /// Does not respect any `PG*` environment variables.
+    ///
+    /// See the type level-documentation for details.
+    pub fn new_without_environment(
+        host: String,
+        port: u16,
+        username: String,
+        ssl_mode: PgSslMode,
```

Comment:
`port` and `ssl_mode` have reasonable defaults: `5432` and `Prefer`, respectively.

`host` could arguably default to `localhost` or the standard Unix socket path for the platform; we already have that logic implemented in this file.

`username` is the hardest to choose a reasonable default for. Defaulting to the OS username doesn't really make sense to me anymore since that assumes the application is running on the same system as the database, which I don't expect is all that common these days; we only default to it because that's what `libpq` does. I would say `postgres` since that's generally the username that the Postgres server is installed under.

Reply:
I mostly did a mechanical transformation of the existing `new_without_pgpass` function. Very happy to go beyond that.

Yeah happy to add a default port. 

But I think we should reconsider on `ssl_mode`, if SSL is supported by the server it behaves like `require` which verifies nothing. Having users explicitly think about `ssl_mode` may be the best option.

Alternatively we go for: 2 functions: `new` which would default to `ssl_mode = Disable` and `new_ssl` with `ssl_mode = verify-full` both documented.

This way we can steer users to the most appropriate defaults and make them explicitly think about what the needs of their environment and its reuqirement is. I'd love to do my part to avoid having to flag (and demo, and report) the MITM-ability of `ssl_mode = require`.

Reply:
> I'd love to do my part to avoid having to flag (and demo, and report) the MITM-ability of ssl_mode = require.

No need, it's well known and documented: https://www.postgresql.org/docs/current/libpq-ssl.html#LIBPQ-SSL-PROTECTION

The existing behavior tries to match that of `libpq` as close as possible because that's what we figured users would expect, and `libpq` maintains its current behavior for backwards-compatibility.

I think this belongs as part of a wider discussion where we talk about choosing more opinionated defaults for `ConnectOptions`, because this applies to MySQL and SQLite as well.